src/libcore/pin.rs

```diff
@@ -390,55 +391,92 @@ use crate::ops::{CoerceUnsized, Deref, DerefMut, DispatchFromDyn, Receiver};
 /// [`Unpin`]: ../../std/marker/trait.Unpin.html
 /// [`pin` module]: ../../std/pin/index.html
 //
-// Note: the derives below, and the explicit `PartialEq` and `PartialOrd`
-// implementations, are allowed because they all only use `&P`, so they cannot move
-// the value behind `pointer`.
+// Note: the `Clone` derive below causes unsoundness as it's possible to implement
+// `Clone` for mutable references.
+// See <https://internals.rust-lang.org/t/unsoundness-in-pin/11311> for more details.
 #[stable(feature = "pin", since = "1.33.0")]
 #[lang = "pin"]
 #[fundamental]
 #[repr(transparent)]
-#[derive(Copy, Clone, Hash, Eq, Ord)]
+#[derive(Copy, Clone)]
 pub struct Pin<P> {
     pointer: P,
 }
 
-#[stable(feature = "pin_partialeq_partialord_impl_applicability", since = "1.34.0")]
+#[stable(feature = "pin_trait_impls", since = "1.41.0")]
 impl<P, Q> PartialEq<Pin<Q>> for Pin<P>
 where
-    P: PartialEq<Q>,
+    P: Deref,
+    P::Target: PartialEq<Q::Target>,
+    Q: Deref,
 {
     fn eq(&self, other: &Pin<Q>) -> bool {
-        self.pointer == other.pointer
+        **self == **other
```

Comment:
This needs a double-`*` because `PartialEq` adds an implicit `&`, right? Might be worth being very explicit about that here.

What about `P::eq(self.deref(), other.deref())` or so?

Reply:
From what I can tell, `**self == **other` is more common in Rust source code, used by `Rc`, `Arc` and `OsString`. However, at the same time, `::eq` is used by `Cow` and `Box`. I also don't see a risk of accidentally using smaller number of stars, as this would cause Rust to warn about infinite recursion.

I'm not sure if I should change that, but if you strongly think this should be changed, it can be changed.

Reply:
Uh, the *inner* `*` actually deref's the `&` in `&self`, and then the outer `*` calls our `deref` impl.
So, the alternative would be `Pin::deref(self) == Pin::deref(other)`. I'll let other reviewers be the judge what they prefer.

Reply:
I'm not sure what's more common in code at large, but I personally **always** write `P::eq` in these sorts of situations (i.e., when I have a wrapper type and I am delegating to the deref target). I consider it best practice since it avoids confusion about how many `*` are needed and also just makes it more obvious what's going on in the first place. I would strongly prefer that here *in particular* since it gives us something to add a comment to, as well. I suspect most people don't do it because they've never thought that much about it, but maybe that's ungenerous of me.

Reply:
This should be resolved now.